### ISSUE TITLE
December 2019 - content update

### ISSUE THREAD

**Issue by DawnmarieDesJardins** [microsoft/MCW-Modernizing-Data-Analytics-with-SQL-Server-2019#5]
This workshop is scheduled for a content update. Please review the current workshop, any open issues, and provide update suggestions for review. Thanks!

**Comment by timahenning** [comment]
Before the Labs
- Looks like there are a few requirements and notes which are a little out of date (e.g., L-series VMs being required, SSMS in use, BDC in preview).
- 4MB of RAM needed for the management system.
- Some requirements might already be pre-installed, particularly curl.
- I'd rewrite Task 3 to use Azure Data Studio and its notebook instead of these notes.  It might be interesting learning how to do it sans notebook but for a workshop, the notebook's going to be the best route to get people started.
- Task 4 could use the notebook results to find these IP addresses, so I could see removing the kubectl command to keep it simple.
- Check into this GitHub issue and see if there's a fix for Python 3.8.

Hands-On Lab
- The architecture image is fine, though Microsoft has a new one on their overview page.  It probably makes sense to swap in the Docs image assuming there are no issues with that.
- Change requirements to match what the Before the Labs has (e.g., probably eliminate SSMS) and bring up to date.  Remove SSMS instructions.
Exercise 1
- Task 1:  ensure that ADS instructions around HDFS still work.  Update any images which are out of date.
- Task 2:  Check on Create External Table from CSV.  I'd write the scripts myself, but sticking with "use the easiest option," the wizard is the easier option.
Exercise 2
- Confirm ADS notebook images are up to date.
- Review each notebook to ensure they still work with the latest version of ADS and see if there are any optimizations to make in them.
Exercise 3
- Review the notebook to ensure that everything is up to date.
- Maybe look at a notebook that uses ML Services against the master node?  That could be out of scope.
Exercise 4
- Task 1:  I kind of wonder how much this fits.  If the curriculum requires it, sure, and that probably necessitates SSMS because ADS doesn't have data classification built in.  Another alternative would be to use ADS + T-SQL scripts and review the sensitivity classification DMV.
- Task 2:  change SSMS to ADS.
Exercise 5
- General:  this seems really tangential to Big Data Clusters.  If it's a necessary part of the curriculum, great, but it feels like it could be replaced with something more BDC-centric like working with the data and storage pools.  Still, I'll continue this by assuming that it stays.
- Task 1:  change SSMS to ADS.
- Tasks 2-4:  Change images to ADS.  The execution plan analysis might change a little bit as a result.
Exercise 6
- Add a section on using Azure Data Studio's Data Virtualization plugin for BDC management.
- Add a section on using ADS troubleshooting notebooks.

**Comment by davel501** [comment]
In task one of "Before the HOL - Modernizing Data Analytics with SQL Server 2019" the link in step #12 is dead (https://docs.microsoft.com/en-us/sql/azure-data-studio/sql-server-2019-extension?view=sql-server-2017).

Also getting the following warning: 
WARNING: You are using pip version 19.2.3, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

Instructions should include an updated version or to ignore the warning.

**Comment by davel501** [comment]
For the cluster creation the example should probably be something like this: "sql2019mcw<your-initials-or-alias>" instead of: "sql2019mcwYOUR_INITIALS_OR_ALIAS". The uppercase letters draw you right to the example so you miss the instructions and do example: sql2019mcwDLEVY to match the example only to fail because of the uppercase letters.

**Comment by davel501** [comment]
Task 4, item 5 in the script bootstrap-sample-db.cmd, line 42 should read "set SQLCMDUSER=admin" rather than sa. I would also recommend changing the task instructions to say look for the ip address with x port number for this and y port number for that as well as saying specifically not to enter the ip addresses.

**Comment by thevinnie** [comment]
> Task 4, item 5 in the script bootstrap-sample-db.cmd, line 42 should read "set SQLCMDUSER=admin" rather than sa. I would also recommend changing the task instructions to say look for the ip address with x port number for this and y port number for that as well as saying specifically not to enter the port numbers.

Agreed on this.  I had to change the sa reference in three places in the bootstrap-sample-db.cmd file.  It's in my notes, but haven't posted yet.

**Comment by davel501** [comment]
HOL step-by step - Modernizing Data Analytics with SQL Server 2019.md > Before the hands-on lab > Connect with Azure Data Studio > Step #3 should be admin instead of sa. I would also reword the password language to remove the default bit. It makes it sound like you do not have to put anything in for a password.

**Comment by davel501** [comment]
HOL step-by step - Modernizing Data Analytics with SQL Server 2019.md > Before the hands-on lab > Connect with SQL Server Management Studio > Step #2 should be admin instead of sa. I would also reword the password language to remove the default bit. It makes it sound like you do not have to put anything in for a password.

**Comment by davel501** [comment]
HOL step-by step - Modernizing Data Analytics with SQL Server 2019.md > Exercise 1: Using data virtualization > Task 1: Create external table from Azure SQL Database - the first 2 objects in the script already exist.

10:18:13 PMStarted executing query at Line 1
Msg 15578, Level 16, State 1, Line 1
There is already a master key in the database. Please drop it before performing this statement.
10:18:13 PMStarted executing query at Line 3
Msg 15530, Level 16, State 1, Line 7
The credential with name "SQLCred" already exists.

**Comment by davel501** [comment]
HOL step-by step - Modernizing Data Analytics with SQL Server 2019.md > Exercise 1: Using data virtualization > Task 2: Create external table from CSV files > Step 10: Changing the table definition to match the screen shot (smallint to tinyint for StockItemID) will result in the following error:

10:38:52 PMStarted executing query at Line 1
Msg 105082, Level 16, State 1, Line 1
105082;Generic ODBC error: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Bulk load data conversion error (overflow) for row 2, column 1 (StockItemID). Additional error <2>: ErrorMsg: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Bulk load data conversion error (overflow) for row 3, column 1 (StockItemID)., SqlState: 42000, NativeError: 4867 Additional error <3>: ErrorMsg: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Bulk load data conversion error (overflow) for row 4, column 1 (StockItemID)., SqlState: 42000, NativeError: 4867 Additional error <4>: ErrorMsg: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Bulk load data conversion error (overflow) for row 5, column 1 (StockItemID)., SqlState: 42000, NativeError: 4867 Additional error <5>: ErrorMsg: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Bulk load data conversion error (overflow) for row 6, column 1 (StockItemID)., SqlState: 42000, NativeError: 4867 .
OLE DB provider "MSOLEDBSQL" for linked server "(null)" returned message "Unspecified error".
Msg 7421, Level 16, State 2, Line 1
Cannot fetch the rowset from OLE DB provider "MSOLEDBSQL" for linked server "(null)". .
Total execution time: 00:00:03.645

**Comment by davel501** [comment]
HOL step-by step - Modernizing Data Analytics with SQL Server 2019.md > Exercise 1: Using data virtualization > Task 1: Create external table from Azure SQL Database - the SQL script creates a table called SQLReviews but that is the only place that name is used, breaking the unified query in Task 3.

**Comment by thevinnie** [comment]
Before the HOL Steps:
	- The "SQL 2019" extension has been deprecated for Azure Data Studio.  There is a new extension called "Data Virtualization" which should be installed.
	- Several of the pip3 uninstall and install commands return an error related tosetuptools:
		○ ERROR: google-auth 1.7.0 has requirement setuptools>=40.3.0, but you'll have setuptools 30.4.0 which is incompatible.
	- The azdata installation gave me several issues.  I tried a few times to get it installed correctly, but I ultimately ended up installing the utility through the MSI installer:
		○ https://aka.ms/azdata-msi
	- Running azdata required me to install Visual Studio Build Tools for C++.  I completed this with instructions in this website:
		○ https://www.scivision.dev/python-windows-visual-c-14-required/
	- Task #3, Step 6 - Prompt for details in the python AKS cluster creation script no longer requires Docker credentials.  That has been commented out in the py file and should be taken out of the documentation
	- In creating the AKS cluster, there were several feedback steps where the cluster was unable to create the service principal.  This did not prevent the cluster from being created, and may be localized to the internal Azure subscription I'm using, but I wanted to call it out.
	- The link for deploy-sql-big-data-aks.py in the documentation fails when clicked on.


HOL:
	- When creating the master key, an error was returned:
		○ Msg 15578, Level 16, State 1, Line 1
		There is already a master key in the database. Please drop it before performing this statement.
	- Images showing dbo.reviews but table was named dbo.SQLReview, this is also present in later scripts
	- In the notebook_03, the create file format fails because it already exists
	- It would be beneficial to clear all results before checking in the notebooks so that the users don't see the results prior to running.

**Comment by davel501** [comment]
Exercise 6 > Task 1 the instructions are bit vague but I think I got it to connect to the right IP. I get a 404 connecting to https://*:30777/portal.

**Comment by davel501** [comment]
Exercise 6 > Task 2 > item 7: second command should be "supervisorctl restart mssql-server"

C:\Users\dlevy>kubectl exec -it master-0  -c mssql-server -n sql2019mcwdlevy -- /bin/bash
root@master-0:/# supervisorctl restart mssql
mssql: ERROR (no such process)
mssql: ERROR (no such process)
root@master-0:/# supervisorctl restart mssql-server
mssql-server: stopped
mssql-server: started
root@master-0:/#

**Comment by davel501** [comment]
Exercise 2 > Task 1 > step 4: I used my existing python install and it failed badly due to a recent update. I was getting server failed to start errors, win32 errors, etc. I did a ctrl+shit+p > "Configure Python for Notebooks" and redid this step to tell it to use a new installation. After the new installation everything worked as expected. I suggest changing this step to always use a new installation.
